### PR TITLE
Refactor facet panel drawing code

### LINKED ISSUE #5482: Feature Request: Break up `FacetX$draw_panels()` into smaller functions
The `draw_panels()` methods are quite beefy with various tasks. Extending facets with custom `draw_panels()` methods is a bit of a pain due to having to repeat/copy a lot of code. As I understand the code, there are three major steps that should be mostly independent of one another:

* Initialise the panels as a gtable.
* Attach axes to panels.
* Attach strips to panels.

There are more granular steps within each task, but I think the above tasks are good chunks to isolate into separate methods to make facets more extensible. My proposal is to keep `draw_panels()` for backward compatibility, but add 3 methods for the tasks above that are called in `draw_panels()`.

### PR DESCRIPTION
**Pull request by teunbrand**
This PR aims to fix #5482.

Briefly, the panel drawing code is now divided over 3 Facet ggproto methods: `init_gtable()`, `attach_axes()` and `attach_strips()`.

My main gripe with the previous code was that it was hard to follow along with what it was doing. I think reasons why it was hard for me to follow are the following:
* The code was alternating between the three tasks, for example `facet_grid()` was first setting up axes, then setting up strips, then initialising the table, then adding the axes and finally adding the strips.
* Some pieces of code were 4 `if` blocks deep which is hard to keep track of.

The primary benefit is that it should now be easier to 'just' tweak the axes, or 'just' tweak the strips in extensions.
Quick benchmark showed no speed difference between current main and this PR.

### CODE REVIEW

**Comment by teunbrand** [comment]
TODO: try unifying `draw_panels()`

**Comment by teunbrand** [comment]
The `draw_panels()` method is now part of the base `Facet` class, as well as the `init_gtable()` method.
`FacetWrap` still has to `ggproto_parent()` the panel drawing code due to interactions with `CoordFlip` that need to be adressed.
I've left `FacetNull` as-is, to keep using its highly simplified method.

File: R/facet-.R [hunk 1]
```diff
@@ -133,8 +133,32 @@ Facet <- ggproto("Facet", NULL,
...
+      cli::cli_abort(
+        "{.fn {snake_class(self)}} can't use free scales with \\
+        {.fn {snake_class(coord)}}."
+      )
+    }
+
+    aspect_ratio <- theme$aspect.ratio
+    if (!is.null(aspect_ratio) && (space$x || space$y)) {
+      cli::cli_abort("Free scales cannot be mixed with a fixed aspect ratio.")
+    }
```
Comment (teunbrand) [inline R/facet-.R:152]: This won't be triggered for FacetWrap as it has no `params$space_free`, which defaults to `list(x = FALSE, y = FALSE)`.

File: R/facet-wrap.R [hunk 2]
```diff
@@ -257,227 +257,203 @@ FacetWrap <- ggproto("FacetWrap", Facet,
...
-    strip_name <- paste0("strip-", substr(params$strip.position, 1, 1))
-    strip_mat <- empty_table
-    strip_mat[panel_pos] <- unlist(unname(strips), recursive = FALSE)[[params$strip.position]]
-    if (params$strip.position %in% c("top", "bottom")) {
-      inside_x <- (theme$strip.placement.x %||% theme$strip.placement %||% "inside") == "inside"
-      if (params$strip.position == "top") {
-        placement <- if (inside_x) -1 else -2
-        strip_pad <- axis_size$top
+  draw_panels = function(self, panels, layout, x_scales, y_scales, ranges, coord, data, theme, params) {
+    if (inherits(coord, "CoordFlip")) {
```
Comment (teunbrand) [inline R/facet-wrap.R:435]: If failed to find a smarter solution to working out this interaction with CoordFlip, so FacetWrap does this first, then calls the parent method.

File: tests/testthat/_snaps/facet-layout.md [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
 # facet_grid throws errors at bad layout specs
 
-    `coord_fixed()` doesn't support free scales.
+    `facet_grid()` can't use free scales with `coord_fixed()`.
```
Comment (teunbrand) [inline tests/testthat/_snaps/facet-layout.md:36]: This wording changed slightly because FacetGrid and FacetWrap share the method that emits this warning

**Review by thomasp85 — APPROVED** [review]
LGTM
